File: src/com/castlefrog/agl/domains/draughts/DraughtsAction.java
package com.castlefrog.agl.domains.draughts;

import com.castlefrog.agl.Action;

public final class DraughtsAction implements Action {
    private int fx;
    private int fy;
    private int tx;
    private int ty;

    private DraughtsAction(int fx, int fy,
                           int tx, int ty) {
        this.fx = fx;
        this.fy = fy;
        this.tx = tx;
        this.ty = ty;
    }

    public static DraughtsAction valueOf(int fx, int fy,
                                         int tx, int ty) {
        return new DraughtsAction(fx, fy, tx, ty);
    }

    public DraughtsAction copy() {
        return this;
    }

    public int getFX() {
        return fx;
    }

    public int getFY() {
        return fy;
    }

    public int getTX() {
        return tx;
    }

    public int getTY() {
        return ty;
    }

    @Override
    public int hashCode() {
        int code = 7 + fx;
        code = 11 * code + fy;
        code = 11 * code + tx;
        code = 11 * code + ty;
        return code;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof DraughtsAction)) {
            return false;
        }
        DraughtsAction action = (DraughtsAction) object;
        return fx == action.getFX() && fy == action.getFY() &&
               tx == action.getTX() && ty == action.getTY();
    }

    @Override
    public String toString() {
        return "(" + fx + "," + fy + "-" + tx + "," + ty + ")";
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/draughts/DraughtsAction.kt
package com.castlefrog.agl.domains.draughts

import com.castlefrog.agl.Action

/**
 * A single step move or jump from one dark square to another.
 * Multi-jumps are played as successive actions while [DraughtsState.mustContinueFromX]
 * forces the same piece to keep capturing. Kings may slide or jump any distance.
 */
data class DraughtsAction(
    val fromX: Int,
    val fromY: Int,
    val toX: Int,
    val toY: Int
) : Action {

    override fun toString(): String = "($fromX,$fromY)->($toX,$toY)"
}
